File: leet-code/2021/0013-roman-to-integer.rs
https://leetcode.com/problems/roman-to-integer/

impl Solution {
    pub fn roman_to_int(s: String) -> i32 {
        let mut result: i32 = 0;
        let mut previous: i32 = 0;
        for c in s.chars() {
            let current = match c {
                'I' => 1,
                'V' => 5,
                'X' => 10,
                'L' => 50,
                'C' => 100,
                'D' => 500,
                'M' => 1000,
                _ => panic!("can't be possible"),
            };
            if previous < current {
                result += current - 2 * previous;
            } else {
                result += current;
            }
            previous = current;
        }
        result
    }
}

File: leet-code/2022/0013-roman-to-integer.cpp
// https://leetcode.com/problems/roman-to-integer/
// Easy

class Solution {
public:
    int romanToInt(string s) {
        std::array<short, 128> map; // no need to initialize (faster)
        map['I'] = 1;
        map['V'] = 5;
        map['X'] = 10;
        map['L'] = 50;
        map['C'] = 100;
        map['D'] = 500;
        map['M'] = 1000;
        int prev = 0, total = 0;
        for (char c : s) {
            short current = map[c];
            total += current;
            if (current > prev) {
                total -= prev*2;
            }
            prev = current;
        }
        return total;
    }
};
